Added functionality for Windows

kannicht opened the pull request:
Hi @gilday,

just did all the steps to ease the PR. I added functionality for Windows.

1. Added method to check on Windows in DarkModeDetector 
2. Added control structure to check either Windows or Mac.
3. Added README.md information.
4. Changed Plugin name to show that plugin isn't Mac exclusive anymore.

Commit:
Added functionality for Windows

kannicht commented:
I added a commit just on reformatting the files as the check initially failed due to that matter.

gilday commented on .idea/misc.xml:
```diff
@@ -4,7 +4,7 @@
     <packageJsonPaths />
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" project-jdk-name="openjdk-1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" project-jdk-name="11" project-jdk-type="JavaSDK">
```
Can we be sure that the versions of IDEA the plugin supports will always use a Java 11 runtime? I've been hesitant to upgrade the plugin to use JDK 11, because I cannot find any documentation from JetBrains that definitively tells me that we can expect IDEA to always use a Java 11 runtime.

kannicht replied:
As far as I know IDEAs future is a Java 11 runtime customized by Jetbrains. They call it JBR11. So I think, since Java 11 is the latest LTS we could think about moving towards this version.

gilday replied:
I did some more digging on this. It looks like JBR11 is not yet guaranteed.

> Starting with 2019.1 version some products are already bundled with JetBrains Runtime version 11 or have the choice between 8 and 11 when downloading. Other products can be switched to JBR11 as described below.

Let's stick with 1.8 until that changes

gilday replied:
Sorry, I meant to include the link where I found that https://intellij-support.jetbrains.com/hc/en-us/articles/206544879-Selecting-the-JDK-version-the-IDE-will-run-under

kannicht replied:
Okay, in this case we should actually stick with Java 8 👍

gilday replied:
JetBrains got back to me about this one and clarified with this article https://blog.jetbrains.com/idea/2020/01/intellij-idea-2020-1-eap/

gilday commented on README.md:
```diff
@@ -24,3 +24,14 @@ command and interpreting its output.
+The plugin polls the Windows Dark Mode configuration by executing the following
+command and interpreting its output.
+
+    HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Themes\Personalize\AppsUseLightTheme
```
> by executing the following command

Is this a command, or a registry key?

kannicht replied:
There is obviously a mistake from my side. One option is to change the phrase quoted by you to "by querying the following registry key" or add "reg query" before the registry key.

kannicht replied:
```suggestion
     reg query HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Themes\Personalize\AppsUseLightTheme
```

kannicht replied:
As stated, this was a problem from my side (classical copy-paste-error). I forgot to change the things.

gilday commented on README.md:
```diff
@@ -24,3 +24,14 @@ command and interpreting its output.
+
+    HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Themes\Personalize\AppsUseLightTheme
+
+When the plugin detects that macOS is using Dark Mode, it sets the IDEA theme to
```
s/macOS/Windows

kannicht replied:
```suggestion
When the plugin detects that Windows is using Dark Mode, it sets the IDEA theme to
```

kannicht replied:
see above.

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeSync.java:
```diff
@@ -50,7 +53,11 @@ public void dispose() {
-    final boolean isDarkMode = isDarkMode();
+    if (SystemInfo.isMacOSMojave) {
+      isDarkMode = isMacOsDarkMode();
+    } else {
```
Let's add a check for Windows with an else that throws if neither macOS Mojave or Windows is detected.

kannicht replied:
I implemented this.

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeSync.java:
```diff
@@ -50,7 +53,11 @@ public void dispose() {
-    final boolean isDarkMode = isDarkMode();
+    if (SystemInfo.isMacOSMojave) {
+      isDarkMode = isMacOsDarkMode();
+    } else {
```
Do all versions of Windows support this registry key / command?

kannicht replied:
Actually I read about an introduction of that key in Windows 8 but a bit more research unveiled that it is a Windows 10 and newer feature. I changed the values according to that.

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeSync.java:
```diff
@@ -23,16 +24,18 @@
-
   private final LafManager lafManager;
   private final DarkModeSyncThemes themes;
+  private boolean isDarkMode;
```
I believe this is only used in the `updateLafIfNecessary` method, so let's make it a local variable there

kannicht replied:
Changed.

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeDetector.java:
```diff
@@ -4,14 +4,16 @@
+/** Queries the environment to see if Dark Mode is enabled */
 final class DarkModeDetector {
 
+  private DarkModeDetector() {}
```
Can you refrain from moving class members within a file if it is not significant to this change? I would like to keep the diff as small as possible for posterity. I suspect this was done by an automated formatter, but there's already an automated formatter configured in the project (google-java-format) and I'd like to just use that formatter to avoid clashing formatters.

kannicht replied:
Your suspicion is correct. I reformatted all files with the wrong settings. I hope, I resolved that issue with the upcoming commit.

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeSync.java:
```diff
@@ -23,16 +24,18 @@
  */
 public final class DarkModeSync implements Disposable {
 
+  private static final Logger logger = Logger.getInstance(DarkModeSync.class);
```
Ditto my concern about moving class members within a file

gilday commented on src/main/resources/META-INF/plugin.xml:
```diff
@@ -1,28 +1,29 @@
 <idea-plugin url="https://github.com/gilday/dark-mode-sync-plugin">
-  <id>com.github.gilday.darkmode</id>
```
Similar to my concern about moving class members within a file, let's avoid changing more than we have to here. I think you've alphabetized the XML elements, but I'd prefer to keep this changeset focused on what is necessary to introduce Windows support

kannicht replied:
see above.

gilday commented on src/test/java/com/github/gilday/darkmode/DarkModeDetectorTest.java:
```diff
@@ -16,6 +16,6 @@ void reads_dark_mode_setting() {
 
     // test passes if isDarkMode completes without error
-    DarkModeDetector.isDarkMode();
+    DarkModeDetector.isMacOsDarkMode();
```
Can you add a similar test for Windows?

kannicht replied:
I implemented a similar test as requested. Be aware, that the test only run on the corresponding OS.

gilday reviewed (changes requested):
Thanks for this contribution. I don't have access to a Windows environment, so I have no way of testing this - I'm relying on you for that. If there's problems with this in the future, I hope you'll be willing to help troubleshoot them, because I will not be able to.

If possible, I'd like to restrict this functionality to the versions of Windows that support Dark Mode. I don't know what those are.

kannicht commented:
I will work on this in the upcoming days. Thank you for your feedback and I'm sorry for any trouble I cause.

gilday commented:
No trouble at all! Just typical code review feedback :-)

Commit:
Reverted to pre-formatting and added changes requested and fixed some bugs.

kannicht commented:
Okay, hopefully everything is alright now. 

I reverted to the pre-reformatted state so all the movements of class members should be undone. 

Additionally I fixed two bugs I discovered during the reverting. 

a) Error in analyzing the reg query (stdout.endswith...)
b) Added Settings-functionality by removing ApplicationConfigurable os-tag

kannicht committed:
Update misc.xml

kannicht committed:
Updated misc.xml due to unintended additions

kannicht commented:
I changed everything to the matters we discussed. :)

gilday commented on README.md:
```diff
@@ -14,13 +14,24 @@ IDEA theme to IntelliJ
+    reg query HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Themes\Personalize\AppsUseLightTheme
+
+When the plugin detects that Windows is using Dark Mode, it sets the IDEA theme to
+the selected theme. When the plugin detects that macOS is not using Dark Mode, it sets
```
"macOS" should be "Windows"

kannicht replied:
Adjusted README.md to totally fit for Mac and Windows (your requested change and I found that I didn't change the general part)

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeDetector.java:
```diff
@@ -4,14 +4,14 @@
-   * @return true if the user has enabled macOS Dark Mode
-   * @throws IllegalStateException when called on not-macOS
+   * @return true if the user has enabled Dark Mode
+   * @throws IllegalStateException when called on Linux
```
but not on Windows? This should probably remain "when called on not-macOS", or we should generalize it to "when fails to read dark mode setting"

kannicht replied:
Changed to: "... when plugin fails to read system dark mode setting. "

Used "system" to clarify that we talk about the system's settings.

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeSync.java:
```diff
@@ -50,7 +52,16 @@ public void dispose() {
   private void updateLafIfNecessary() {
     final LookAndFeelInfo current = lafManager.getCurrentLookAndFeel();
-    final boolean isDarkMode = isDarkMode();
+    boolean isDarkMode;
```
Let's make this `final` to indicate that it must be set by the following conditional

kannicht replied:
Done.

gilday reviewed (changes requested):
Nearly there!

Commit:
Changes as requested in PR.

Commit:
Typo corrected.

kannicht commented:
Still getting closer:

All changes are explained in the corresponding change requests.

Checks might fail due to wrong formatting: but I learned from my past mistakes ;)

gilday commented:
Great! The checks are indeed failing due to formatting. If you simply run `./gradlew goJF`, the formatter will fix any formatting issues, then we can get this merged.

Commit:
Reformatted using Gradle.

kannicht commented:
So 😄

Checks passed, ready to go 👍
